cpp/include/cudf/detail/copy_if.cuh

```diff
@@ -0,0 +1,412 @@
+                                          cudf::size_type &block_sum)
+{
+  int offset = 0;
+
+  using BlockScan = cub::BlockScan<cudf::size_type, block_size>;
+  __shared__ typename BlockScan::TempStorage temp_storage;
+  BlockScan(temp_storage).ExclusiveSum(mask_true, offset, block_sum);
+
+  return offset;
+}
+
+template <typename Kernel>
+int elements_per_thread(Kernel kernel,
+                        cudf::size_type total_size,
+                        cudf::size_type block_size)
+{
+  // calculate theoretical occupancy
+  int max_blocks = 0;
+  CUDA_TRY(cudaOccupancyMaxActiveBlocksPerMultiprocessor(&max_blocks, kernel,
+                                                         block_size, 0));
+
+  int device = 0;
+  CUDA_TRY(cudaGetDevice(&device));
+  int num_sms = 0;
+  CUDA_TRY(cudaDeviceGetAttribute(&num_sms, cudaDevAttrMultiProcessorCount, device));
+  int per_thread = total_size / (max_blocks * num_sms * block_size);
+  return std::max(1, std::min(per_thread, 32)); // switch to std::clamp with C++17
+}
+
+// Computes the output size of apply_boolean_mask, which is the sum of the
+// last block's offset and the last block's pass count
+cudf::size_type get_output_size(cudf::size_type * block_counts,
+                              cudf::size_type * block_offsets,
+                              cudf::size_type num_blocks,
+                              cudaStream_t stream = 0)
+{
+  cudf::size_type last_block_count = 0;
+  CUDA_TRY(cudaMemcpyAsync(&last_block_count, &block_counts[num_blocks - 1],
+                  sizeof(cudf::size_type), cudaMemcpyDefault, stream));
+  cudf::size_type last_block_offset = 0;
+  if (num_blocks > 1)
+    CUDA_TRY(cudaMemcpyAsync(&last_block_offset, &block_offsets[num_blocks - 1],
+                    sizeof(cudf::size_type), cudaMemcpyDefault, stream));
+  cudaStreamSynchronize(stream);
+  return last_block_count + last_block_offset;
+}
+
+// This kernel scatters data and validity mask of a column based on the
+// scan of the boolean mask. The block offsets for the scan are already computed.
+// Just compute the scan of the mask in each block and add it to the block's
+// output offset. This is the output index of each element. Scattering
+// the valid mask is not as easy, because each thread is only responsible for
+// one bit. Warp-level processing (ballot) makes this simpler.
+// To make scattering efficient, we "coalesce" the block's scattered data and
+// valids in shared memory, and then write from shared memory to global memory
+// in a contiguous manner.
+// The has_validity template parameter specializes this kernel for the
+// non-nullable case for performance without writing another kernel.
+//
+// Note: `filter` is not run on indices larger than the input column size
+template <typename T, typename Filter,
+          int block_size, bool has_validity>
+__launch_bounds__(block_size)
+__global__ void scatter_kernel(T* __restrict__ output_data,
```

Comment:
@harrism @rgsl888prabhu while this kernel will certainly be more efficient for fixed-width types, we could get string support "for free" by instead modifying `copy_if` to use `cudf::scatter`, which will already have string support.

Reply:
Is scatter ready for use ?

Reply:
Not yet no, but shouldn't be too long now: https://github.com/rapidsai/cudf/pull/3354

Reply:
@harrism do you have any suggestion regarding this ?

Reply:
And on second thought, I think `copy_if` is a more natural fit for `cudf::gather` than `cudf::scatter`.

Reply:
Suggestion: run `apply_boolean_mask_benchmark` on a V100 (on a DGX) to get a baseline before this PR. Save the results. (I can run this if you want, or you can.) Try just calling scatter and see how it compares. The reason I originally wrote a custom kernel here was because the "scatter" is more limited -- it's monotonically increasing indices -- which means it can be optimized for that case. However, if generic scatter is close enough to the performance in the benchmark, it's worth simplifying.

Reply:
I will check the difference between  both and will get back.

Reply:
Looks like this kernel is 50-70% faster for fixed-width types, which is not surprising -- there's a reason it's a custom kernel. :)  I suggest keeping it to avoid boolean masking regressions  in cuDF. We special case to use gather for strings columns.